### PR TITLE
Windows

### PR DESCRIPTION
**Pull request by MisterDA**
This pull request:
- updates vendored dependencies to their latest commit (on master branch) (required for client bug-fixes);
- updates the minimum OCaml version to 4.12;
- updates the requirement on Lwt to 5.4.1 for Windows bugfixes;
- adds a pipeline for building Windows Docker images containing fdopen's Docker for Windows Opam and OCaml installations;
- allows to run the client on Windows.

Building Windows Docker images requires OCluster workers running on Windows (see OCluster PR 128) are used. The current pipeline doesn't create multiarch images for different OS (i.e., an image containing Linux and Windows images). It doesn't change the current Linux tags in any way.

It uses three new distros: _windows-mingw_, _windows-msvc_, and _cygwin_, corresponding to the three OCaml Windows ports. Cygwin is currently disabled until newer versions of OCaml (4.12.1 ?) and Opam (2.1 ?) are made available.

The pushed Windows tags have the current scheme, where `<distro>` is one of `windows-mingw` or `windows-msvc`:
- `<distro>-<win_version>-opam`;
- `<distro>-<win_version>-ocaml-<ocaml_version>`;
- `<distro>-<win_version>` -> `<distro>-<win_version>-ocaml-<ocaml_version_latest>`;
- `<distro>-ocaml-<ocaml_version>`, a multiarch image containing all `<distro>-<win_version>-ocaml-<ocaml_version>` for each `win_version`.
- `<distro>` , a multiarch image containing all `<distro>-<win_version>` for each `<win_version>`.

The `<win_version>` corresponds to the `mcr.microsoft.com/windows:<win_version>` base image.

Some possible set of tags that aren't in this PR is:

- `<distro>-opam`, a multiarch image containing all `<distro>-<win_version>-opam` for all `win_version`, but as opam images are intermediary, we don't have to bother with it.

A image is generated for each version of Windows 10 (LTSC and Semi Annual releases only). Images for earlier versions of Windows 10 are not available. As Microsoft doesn't provide a "latest" tags for its images, we won't either provide an image based on the "latest" version of Windows.

Depends on :
- [x] https://github.com/avsm/ocaml-dockerfile/pull/32
- [x] https://github.com/avsm/ocaml-dockerfile/pull/41
- [x] https://github.com/avsm/ocaml-dockerfile/pull/42
- [x] https://github.com/avsm/ocaml-dockerfile/pull/50
- [x] https://github.com/avsm/ocaml-dockerfile/pull/53
- [x] https://github.com/avsm/ocaml-dockerfile/pull/54
- [x] https://github.com/ocurrent/ocurrent/pull/270
- [x] https://github.com/ocurrent/ocurrent/pull/268
- [x] https://github.com/ocurrent/ocurrent/pull/253
- [x] https://github.com/ocurrent/ocurrent/pull/246
- [x] https://github.com/ocurrent/ocurrent/pull/233
- [x] https://github.com/ocurrent/docker-base-images/pull/119
- [x] https://github.com/ocurrent/docker-base-images/pull/120
- [x] Lwt 5.4.1 + patch

Might depend on (but not sure as I haven't changed the API used):
- [ ] https://github.com/ocurrent/ocluster/pull/128

### CODE REVIEW

**Review by talex5 — APPROVED** [review]

**Comment by avsm** [comment]
All the prerequisites for this are now merged.

**Comment by talex5** [comment]
CI is failing. Probably the Dockerfile needs updating.

**Comment by MisterDA** [comment]
The CI's passing!
The new code depends on ocaml-dockerfile#41 though.

File: src/pipeline.ml
Comment (talex5) [inline src/pipeline.ml:97]:
```suggestion
  entrypoint_exec (Option.to_list personality @ etp_exec) @@
```

File: src/pipeline.ml
Comment (talex5) [inline src/pipeline.ml:107]: Would be good to add a comment saying why we only squash on Linux.

**Review by talex5 — COMMENTED** [review]
This looks reasonable (I can't check the Windows bits), but needs rebasing. What needs to happen to get this deployed? I assume we need to at least add some Windows workers to the main cluster.

**Comment by MisterDA** [comment]
I can't think of new changes to bring to docker-base-images, but I still have some work on the dependencies.
I have a few minor updates to make on ocaml-dockerfile as Cygwin was just updated. Regarding the deployment, I need to add support for the ocluster workers and schedulers to be launched as Windows services, that's a bit tricky. I also need to ping the ExtUnix maintainer to tag a new release with the bug fixes, otherwise the workers and schedulers will disconnect.
I'll apply your suggestions (thanks!) and rebase, and ping back when I'm confident in the changes.

File: src/pipeline.ml
Comment (MisterDA) [reply src/pipeline.ml:97]: Applied, thanks!

File: src/pipeline.ml
Comment (MisterDA) [reply src/pipeline.ml:106]: Done!

File: src/conf.ml
Comment (dra27) [inline src/conf.ml:8]: Why's this in `Docker`? I'd take the root as `"/run/secrets"` on Unix and `Option.value (Sys.getenv_opt "ProgramData") ~default:"C:\ProgramData")` on Windows and then put the secrets in `ocurrent-hub` under that.

File: src/conf.ml
Comment (dra27) [inline src/conf.ml:43]: I think the reason we're doing this is because opam-repository-mingw doesn't have the variants we want? In which case, please put a comment for that 🙂

File: src/pipeline.ml
Comment (dra27) [reply src/pipeline.ml:102]: 2021...!

File: src/git_repositories.ml
Comment (dra27) [inline src/git_repositories.ml:19]: At some point later this year this will go - might it be better to have a list of repositories here?

File: src/pipeline.ml
Comment (dra27) [inline src/pipeline.ml:57]: `etp_exec` is a bit cryptic - just `entrypoint`?

File: src/pipeline.ml
Comment (dra27) [inline src/pipeline.ml:324]: FIXME note

File: src/pipeline.ml
Comment (dra27) [inline src/pipeline.ml:310]: I haven't stared too hard, but is there not some common code with the linux pipeline above which could be factored out now?

**Review by dra27 — COMMENTED** [review]
Some possible refactorings, but this is basically looking marvellous!

File: src/git_repositories.ml
Comment (MisterDA) [reply src/git_repositories.ml:19]: I think it doesn't matter much, if the patch is sufficiently clean it can simply be reverted, but maybe a list would help managing multiple repositories. Ok!

Comment (MisterDA) [reply src/git_repositories.ml:19]: Hm, no, it's simpler with a flat structure (no list), because the branch we want to checkout is different for each git repo (master for opam-repository, opam2 for opam-repository-mingw). With a list we won't be able to distinguish the repos, so we'd need to attach the branch to clone. It would induce a somewhat complex refactor, while reverting the commit is simple enough to remove the support for opam-repository-mingw.

File: src/pipeline.ml
Comment (MisterDA) [reply src/pipeline.ml]: Changed to `opam_exec`, `entrypoint` would shadow `Dockerfile.entrypoint`.

File: src/pipeline.ml
Comment (MisterDA) [reply src/pipeline.ml]: Removed the note, as I haven't seen any duplicates.

File: src/pipeline.ml
Comment (MisterDA) [reply src/pipeline.ml]: There was, but I'm not sure if it improves readability substantially -- might improve maintainance as expected.

File: src/pipeline.ml
Comment (avsm) [inline src/pipeline.ml:101]: I think this isn't right, since the base-images _host_ will be Linux and so Sys.unix will be always set. It needs to tell the system it's dispatching to is a Linux or Windows distro and decide like that.

Comment (MisterDA) [reply src/pipeline.ml]: Well spotted, thank you! Changed to use the OS family of the current distro.

**Comment by talex5** [comment]
I split out some commits to #119 and #120. #119 is merged and I rebased on that. If you're happy with #120 we can merge and rebase on that too. That should leave just Windows stuff in this PR.

File: src/conf.ml
Comment (MisterDA) [reply src/conf.ml:8]: @dra27 we haven't really resolved that conversation, are you convinced by my explanation?

**Comment by avsm** [comment]
Winget Pr merged into dockerfile

File: builds.expected
Comment (talex5) [inline builds.expected:120]: Was this change to the order of the commands intentional?

File: builds.expected
Comment (talex5) [inline builds.expected:124]: This shouldn't be removed.

**Review by talex5 — COMMENTED** [review]
Thanks, this is looking very good!

I rebased it on master to get the new diff test. There are two changes to the Linux builds that are probably accidental. I'm not qualified to say whether the Windows builds work.

Things to do before merging:

- [ ] Undo the changes to the Linux builds.
- [ ] @MisterDA mentioned that ocaml-dockerfile should be updated to remove the `[TMP]` bit in the last commit).

**Comment by talex5** [comment]
Deployed live, but some of the builds failed.

https://base-images.ocamllabs.io/job/2021-07-08/165647-ocluster-build-fde59f:
```
#=== ERROR while compiling ocaml-variants.4.09.1+mingw64 ======================#
# /bin/dash: 1: iconv: not found
# make[4]: *** [../../Makefile.common:71: getserv.o] Error 1
```

https://base-images.ocamllabs.io/job/2021-07-08/165647-ocluster-build-65fbee:
```
[ERROR] The sources of the following couldn't be obtained, aborting:
          - config-file.1.2: curl error code       0 [main] curl 834 child_info_fork::abort: \??\C:\cygwin64\bin\cygssl-1.1.dll: Loaded to different address: parent(0x3FE3A0000) != child(0x400000)
          - depext-cygwinports.0.0.8: curl error code       0 [main] curl 835 child_info_fork::abort: \??\C:\cygwin64\bin\cygssl-1.1.dll: Loaded to different address: parent(0x3FE3A0000) != child(0x400000)
```

https://base-images.ocamllabs.io/job/2021-07-08/165647-ocluster-build-dfdae0:
```
# File "C:\opam\.opam\4.06\.opam-switch\build\ocaml-variants.4.06.1+mingw64\middle_end\inlining_decision.ml", line 1:
# Error: Assembler error, input left in file C:\cygwin64\tmp\camlasmc2b619.s
# make[1]: *** [Makefile:1297: middle_end/inlining_decision.cmx] Error 2
```

**Comment by talex5** [comment]
After clicking the rebuild buttons a few times, everything is now passing.

**Pull request merged**
